src/flags.zig

```diff
@@ -302,84 +302,136 @@ fn parse_value(comptime T: type, flag: []const u8, value: [:0]const u8) T {
-        const suffixes = unit[0];
-        const scale = unit[1];
-        for (suffixes) |suffix| {
-            if (std.mem.endsWith(u8, value, suffix)) {
-                break :unit .{ .suffix = suffix, .scale = scale };
-            }
-        }
-    } else break :unit .{ .suffix = "", .scale = 1 };
-
-    assert(std.mem.endsWith(u8, value, unit.suffix));
-    const value_numeric = value[0 .. value.len - unit.suffix.len];
-
-    const amount = std.fmt.parseUnsigned(u64, value_numeric, 10) catch |err| {
+    return ByteSize.parse(value) catch |err| {
         switch (err) {
-            error.Overflow => fatal(
+            error.ParseOverflow => fatal(
                 "{s}: value exceeds 64-bit unsigned integer: '{s}'",
                 .{ flag, value },
             ),
             error.InvalidCharacter => fatal(
                 "{s}: expected a size, but found '{s}' (invalid digit or suffix)",
                 .{ flag, value },
             ),
+            error.BytesOverflow => fatal(
+                "{s}: size in bytes exceeds 64-bit unsigned integer: '{s}'",
+                .{ flag, value },
+            ),
         }
     };
-
-    const bytes = std.math.mul(u64, amount, unit.scale) catch fatal(
-        "{s}: size in bytes exceeds 64-bit unsigned integer: '{s}'",
-        .{ flag, value },
-    );
-    return ByteSize{ .bytes = bytes };
 }
 
+pub const ByteUnit = enum(u64) {
+    bytes = 1,
+    kib = 1024,
+    mib = 1024 * 1024,
+    gib = 1024 * 1024 * 1024,
+    tib = 1024 * 1024 * 1024 * 1024,
+};
+
+const ByteSizeParseError = error{
+    ParseOverflow,
+    InvalidCharacter,
+    BytesOverflow,
+};
+
+pub const ByteSize = struct {
+    value: u64,
+    unit: ByteUnit = .bytes,
+
+    fn parse(value: []const u8) ByteSizeParseError!ByteSize {
+        const units = .{
+            .{ &[_][]const u8{ "TiB", "tib" }, ByteUnit.tib },
+            .{ &[_][]const u8{ "GiB", "gib" }, ByteUnit.gib },
+            .{ &[_][]const u8{ "MiB", "mib" }, ByteUnit.mib },
+            .{ &[_][]const u8{ "KiB", "kib" }, ByteUnit.kib },
+        };
+
+        const unit: struct {
+            value_input: []const u8,
+            unit: ByteUnit,
+        } = unit: inline for (units) |unit| {
+            const suffixes = unit[0];
+            const unit_kind = unit[1];
+            for (suffixes) |suffix_| {
+                if (std.mem.endsWith(u8, value, suffix_)) {
+                    break :unit .{
+                        .value_input = value[0 .. value.len - suffix_.len],
+                        .unit = unit_kind,
+                    };
+                }
+            }
+        } else break :unit .{ .value_input = value, .unit = ByteUnit.bytes };
+
+        const amount = std.fmt.parseUnsigned(u64, unit.value_input, 10) catch |err| {
+            switch (err) {
+                error.Overflow => {
+                    return ByteSizeParseError.ParseOverflow;
+                },
+                error.InvalidCharacter => {
```

Comment:
As a colateral effect of not accepting MB or GB anymore, this error feels terribly cryptic now (imagine asking for 1gb and seeing this 😬)

Could we include MB/GB/TB only for the purpose of returning an specific error telling the user to use MiB/GiB/TiB instead?

Reply:
Or maybe a simpler solution: just replace the error message with something like "invalid unit, it should be MiB, GiB or TiB"

Reply:
I've addressed this now. I restructured this parser a bit to distinguish between incorrect units and incorrect digits.